File: customgpt-widget-next/src/lib/ai/customgpt-client.ts
/**
 * CustomGPT API Client
 *
 * Handles all communication with CustomGPT Conversations API.
 * Supports conversation creation, message sending, and streaming responses.
 */

import { CUSTOMGPT_CONFIG, LANGUAGE_CONFIG } from '@/config/constants';

const BASE_URL = CUSTOMGPT_CONFIG.apiBaseUrl;
const PROJECT_ID = CUSTOMGPT_CONFIG.projectId;
const API_KEY = CUSTOMGPT_CONFIG.apiKey;
const LANGUAGE = LANGUAGE_CONFIG.default;

if (!PROJECT_ID) {
  throw new Error('CUSTOMGPT_PROJECT_ID environment variable is required');
}
if (!API_KEY) {
  throw new Error('CUSTOMGPT_API_KEY environment variable is required');
}

/**
 * Response types
 */
export interface ConversationData {
  id: number;
  session_id: string;
  project_id: number;
  created_at: string;
}

export interface MessageData {
  id: number;
  user_query: string;
  openai_response: string;
  citations?: number[];
  created_at: string;
  response_feedback?: {
    created_at: string;
    updated_at: string;
    user_id: number;
    reaction: 'liked' | 'disliked' | null;
  };
}

export interface ApiResponse<T> {
  status: 'success' | 'error';
  data: T;
  message?: string;
}

export interface StreamData {
  status: 'progress' | 'finish' | 'error';
  message?: string;
  error?: string;
}

export interface AgentSettings {
  chatbot_avatar: string | null;
  chatbot_background_type?: string;
  chatbot_background?: string;
  chatbot_background_color?: string;
  default_prompt?: string;
  example_questions: string[];
  response_source?: string;
  chatbot_msg_lang?: string;
  chatbot_color?: string;
  chatbot_toolbar_color?: string;
  persona_instructions?: string;
  citations_answer_source_label_msg?: string;
  citations_sources_label_msg?: string;
  hang_in_there_msg?: string;
  chatbot_siesta_msg?: string;
  is_loading_indicator_enabled?: boolean;
  enable_citations?: number;
  enable_feedbacks?: boolean;
  citations_view_type?: string;
  image_citation_display?: string;
  no_answer_message?: string;
  ending_message?: string;
  try_asking_questions_msg?: string;
  view_more_msg?: string;
  view_less_msg?: string;
  remove_branding?: boolean;
  private_deployment?: boolean;
  enable_recaptcha_for_public_chatbots?: boolean;
  chatbot_model?: string;
  is_selling_enabled?: boolean;
  license_slug?: boolean;
  selling_url?: string;
  can_share_conversation?: boolean;
  can_export_conversation?: boolean;
  hide_sources_from_responses?: boolean;
  input_field_addendum?: string;
  user_avatar?: string;
  spotlight_avatar_enabled?: boolean;
  spotlight_avatar?: string;
  spotlight_avatar_shape?: string;
  spotlight_avatar_type?: string;
  user_avatar_orientation?: string;
  chatbot_title: string;
  chatbot_title_color?: string;
  enable_inline_citations_api?: boolean;
  conversation_time_window?: boolean;
  conversation_retention_period?: string;
  conversation_retention_days?: number;
  enable_agent_knowledge_base_awareness?: boolean;
  markdown_enabled?: boolean;
}

export interface AgentDetails {
  id: number;
  project_name: string;
  sitemap_path?: string;
  is_chat_active: boolean;
  user_id: number;
  team_id: number;
  created_at: string;
  updated_at: string;
  deleted_at?: string | null;
  type: string;
  is_shared: boolean;
  shareable_slug?: string;
  shareable_link?: string;
  embed_code?: string;
  live_chat_code?: string;
  are_licenses_allowed?: boolean;
}

/**
 * CustomGPT API Client
 */
export class CustomGPTClient {
  private baseUrl: string;
  private projectId: string;
  private apiKey: string;
  private language: string;

  constructor() {
    this.baseUrl = BASE_URL;
    this.projectId = PROJECT_ID!;
    this.apiKey = API_KEY!;
    this.language = LANGUAGE;
  }

  /**
   * Get headers for API requests
   */
  private getHeaders(): HeadersInit {
    return {
      'accept': 'application/json',
      'content-type': 'application/json',
      'authorization': `Bearer ${this.apiKey}`,
    };
  }

  /**
   * Create a new conversation
   *
   * @returns Conversation data with session_id
   */
  async createConversation(): Promise<ConversationData> {
    const url = `${this.baseUrl}/projects/${this.projectId}/conversations`;

    // CustomGPT API requires a "name" field
    const payload = { name: 'Chat Conversation' };

    const response = await fetch(url, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(`Failed to create conversation: ${response.status} ${response.statusText}`);
    }

    const data: ApiResponse<ConversationData> = await response.json();

    if (data.status !== 'success') {
      throw new Error(`Failed to create conversation: ${data.message || 'Unknown error'}`);
    }

    return data.data;
  }

  /**
   * Send a message to a conversation (non-streaming)
   *
   * @param sessionId - The conversation session ID
   * @param userMessage - The user's message text
   * @returns Message response with AI response and citations
   */
  async sendMessage(sessionId: string, userMessage: string): Promise<MessageData> {
    const url = `${this.baseUrl}/projects/${this.projectId}/conversations/${sessionId}/messages`;

    const params = new URLSearchParams({
      stream: 'false',
      lang: this.language,
    });

    const payload = {
      prompt: userMessage,
      response_source: 'default',
    };

    const response = await fetch(`${url}?${params}`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(`Failed to send message: ${response.status} ${response.statusText}`);
    }

    const data: ApiResponse<MessageData> = await response.json();

    if (data.status !== 'success') {
      throw new Error(`Failed to send message: ${data.message || 'Unknown error'}`);
    }

    return data.data;
  }

  /**
   * Send a message and stream the response using Server-Sent Events
   *
   * @param sessionId - The conversation session ID
   * @param userMessage - The user's message text
   * @returns AsyncGenerator yielding chunks of the AI response
   */
  async *sendMessageStream(sessionId: string, userMessage: string): AsyncGenerator<string, void, unknown> {
    const url = `${this.baseUrl}/projects/${this.projectId}/conversations/${sessionId}/messages`;

    const params = new URLSearchParams({
      stream: 'true',
      lang: this.language,
    });

    const payload = {
      prompt: userMessage,
      response_source: 'default',
    };

    const response = await fetch(`${url}?${params}`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(`Failed to send message: ${response.status} ${response.statusText}`);
    }

    if (!response.body) {
      throw new Error('Response body is null');
    }

    // Parse SSE stream
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');

        // Keep the last incomplete line in buffer
        buffer = lines.pop() || '';

        for (const line of lines) {
          if (!line.trim()) continue;

          // SSE format: "data: {json}"
          if (line.startsWith('data: ')) {
            const dataStr = line.substring(6);

            try {
              const data: StreamData = JSON.parse(dataStr);

              // Handle progress events with message chunks
              if (data.status === 'progress' && data.message) {
                yield data.message;
              }

              // Handle finish event (end of stream)
              if (data.status === 'finish') {
                return;
              }

              // Handle error event
              if (data.status === 'error') {
                throw new Error(data.error || 'Stream error occurred');
              }
            } catch (error) {
              if (error instanceof SyntaxError) {
                // Skip malformed JSON
                continue;
              }
              throw error;
            }
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  /**
   * Get all messages in a conversation
   *
   * @param sessionId - The conversation session ID
   * @returns List of messages in the conversation
   */
  async getConversationMessages(sessionId: string): Promise<MessageData[]> {
    const url = `${this.baseUrl}/projects/${this.projectId}/conversations/${sessionId}/messages`;

    const response = await fetch(url, {
      method: 'GET',
      headers: this.getHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to get messages: ${response.status} ${response.statusText}`);
    }

    const data: ApiResponse<MessageData[]> = await response.json();

    if (data.status !== 'success') {
      throw new Error(`Failed to get messages: ${data.message || 'Unknown error'}`);
    }

    return data.data;
  }

  /**
   * Update reaction for a specific message
   *
   * @param sessionId - The conversation session ID
   * @param messageId - The message ID (prompt_id)
   * @param reaction - "liked", "disliked", or null to remove reaction
   * @returns Updated message data with response_feedback
   */
  async updateMessageReaction(
    sessionId: string,
    messageId: number,
    reaction: 'liked' | 'disliked' | null
  ): Promise<MessageData> {
    // Validate reaction value
    if (reaction !== 'liked' && reaction !== 'disliked' && reaction !== null) {
      throw new Error(`Invalid reaction value: ${reaction}. Must be 'liked', 'disliked', or null`);
    }

    const url = `${this.baseUrl}/projects/${this.projectId}/conversations/${sessionId}/messages/${messageId}/feedback`;

    const payload = { reaction };

    console.log('[CustomGPT] Updating message reaction:', { url, payload });

    const response = await fetch(url, {
      method: 'PUT',
      headers: this.getHeaders(),
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      // Try to get error details from response body
      let errorDetails = `${response.status} ${response.statusText}`;
      try {
        const errorBody = await response.json();
        errorDetails = errorBody.data?.message || errorBody.message || errorDetails;
        console.error('[CustomGPT] Feedback API error response:', errorBody);
      } catch {
        // Couldn't parse error body
      }
      throw new Error(`Failed to update reaction: ${errorDetails}`);
    }

    const data: ApiResponse<MessageData> = await response.json();

    if (data.status !== 'success') {
      throw new Error(`Failed to update reaction: ${data.message || 'Unknown error'}`);
    }

    console.log('[CustomGPT] Reaction updated successfully');
    return data.data;
  }

  /**
   * Get citation details
   *
   * @param citationId - The citation ID
   * @returns Citation details
   */
  async getCitationDetails(citationId: number): Promise<any> {
    const url = `${this.baseUrl}/projects/${this.projectId}/citations/${citationId}`;

    const response = await fetch(url, {
      method: 'GET',
      headers: this.getHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to get citation: ${response.status} ${response.statusText}`);
    }

    const data: ApiResponse<any> = await response.json();

    if (data.status !== 'success') {
      throw new Error(`Failed to get citation: ${data.message || 'Unknown error'}`);
    }

    return data.data;
  }

  /**
   * Get agent settings
   *
   * @returns Agent settings including title, avatar, example questions, etc.
   */
  async getAgentSettings(): Promise<AgentSettings> {
    const url = `${this.baseUrl}/projects/${this.projectId}/settings`;

    const response = await fetch(url, {
      method: 'GET',
      headers: this.getHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to get agent settings: ${response.status} ${response.statusText}`);
    }

    const data: ApiResponse<AgentSettings> = await response.json();

    if (data.status !== 'success') {
      throw new Error(`Failed to get agent settings: ${data.message || 'Unknown error'}`);
    }

    return data.data;
  }

  /**
   * Get agent details
   *
   * @returns Agent details including name, type, status, etc.
   */
  async getAgentDetails(): Promise<AgentDetails> {
    const url = `${this.baseUrl}/projects/${this.projectId}`;

    const response = await fetch(url, {
      method: 'GET',
      headers: this.getHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to get agent details: ${response.status} ${response.statusText}`);
    }

    const data: ApiResponse<AgentDetails> = await response.json();

    if (data.status !== 'success') {
      throw new Error(`Failed to get agent details: ${data.message || 'Unknown error'}`);
    }

    return data.data;
  }
}

// Export singleton instance
export const customGPTClient = new CustomGPTClient();

File: customgpt-widget/backend/customgpt_client.py
"""
CustomGPT API Client

Handles all communication with CustomGPT Conversations API.
Supports conversation creation, message sending, and streaming responses.
"""

import os
import json
import requests
import httpx
from typing import Optional, Dict, Any, AsyncGenerator
from dotenv import load_dotenv

load_dotenv()

class CustomGPTClient:
    """Client for CustomGPT API interactions"""

    def __init__(self):
        self.base_url = "https://app.customgpt.ai/api/v1"
        self.project_id = os.getenv("CUSTOMGPT_PROJECT_ID")
        self.api_key = os.getenv("CUSTOMGPT_API_KEY")
        self.language = os.getenv("LANGUAGE", "en")

        if not self.project_id:
            raise ValueError("CUSTOMGPT_PROJECT_ID environment variable is required")
        if not self.api_key:
            raise ValueError("CUSTOMGPT_API_KEY environment variable is required")

    def _get_headers(self) -> Dict[str, str]:
        """Get headers for API requests"""
        return {
            "accept": "application/json",
            "content-type": "application/json",
            "authorization": f"Bearer {self.api_key}"
        }

    def create_conversation(self) -> Dict[str, Any]:
        """
        Create a new conversation.

        Returns:
            dict: Conversation data with session_id

        Example response:
        {
            "status": "success",
            "data": {
                "id": 1,
                "session_id": "f1b9aaf0-5e4e-11eb-ae93-0242ac130002",
                "project_id": 123,
                "created_at": "2023-04-30 16:43:53"
            }
        }
        """
        url = f"{self.base_url}/projects/{self.project_id}/conversations"

        # CustomGPT API requires a "name" field
        payload = {"name": "Chat Conversation"}

        response = requests.post(url, headers=self._get_headers(), json=payload)
        response.raise_for_status()

        data = response.json()
        if data.get("status") != "success":
            raise Exception(f"Failed to create conversation: {data}")

        return data["data"]

    def send_message(
        self,
        session_id: str,
        user_message: str,
        stream: bool = False
    ) -> Dict[str, Any]:
        """
        Send a message to a conversation.

        Args:
            session_id: The conversation session ID
            user_message: The user's message text
            stream: Whether to stream the response

        Returns:
            dict: Message response with AI response and citations

        Example response:
        {
            "status": "success",
            "data": {
                "id": 1,
                "user_query": "Hello",
                "openai_response": "Hi! How can I help?",
                "citations": [1, 2],
                "created_at": "2021-01-01 00:00:00"
            }
        }
        """
        url = f"{self.base_url}/projects/{self.project_id}/conversations/{session_id}/messages"

        params = {
            "stream": "true" if stream else "false",
            "lang": self.language
        }

        payload = {
            "prompt": user_message,
            "response_source": "default"
        }

        response = requests.post(
            url,
            headers=self._get_headers(),
            params=params,
            json=payload
        )
        response.raise_for_status()

        data = response.json()
        if data.get("status") != "success":
            raise Exception(f"Failed to send message: {data}")

        return data["data"]

    async def send_message_stream(
        self,
        session_id: str,
        user_message: str
    ) -> AsyncGenerator[str, None]:
        """
        Send a message and stream the response using Server-Sent Events.

        Args:
            session_id: The conversation session ID
            user_message: The user's message text

        Yields:
            str: Chunks of the AI response as they arrive
        """
        url = f"{self.base_url}/projects/{self.project_id}/conversations/{session_id}/messages"

        params = {
            "stream": "true",
            "lang": self.language
        }

        payload = {
            "prompt": user_message,
            "response_source": "default"
        }

        # Using httpx for async streaming
        async with httpx.AsyncClient(timeout=60.0) as client:
            async with client.stream(
                "POST",
                url,
                headers=self._get_headers(),
                params=params,
                json=payload
            ) as response:
                response.raise_for_status()

                # Stream the response line by line
                # CustomGPT SSE format:
                # event: progress
                # data: {"status":"progress","message":"Hello"}
                async for line in response.aiter_lines():
                    if line:
                        # SSE format: "data: {json}"
                        if line.startswith('data: '):
                            data_str = line[6:]  # Remove "data: " prefix

                            try:
                                data = json.loads(data_str)

                                # CustomGPT streaming format
                                # Handle progress events with message chunks
                                if data.get('status') == 'progress':
                                    message = data.get('message', '')
                                    if message:
                                        yield message

                                # Handle finish event (end of stream)
                                elif data.get('status') == 'finish':
                                    break

                            except json.JSONDecodeError:
                                # Skip malformed JSON
                                continue

    def get_conversation_messages(self, session_id: str) -> list:
        """
        Get all messages in a conversation.

        Args:
            session_id: The conversation session ID

        Returns:
            list: List of messages in the conversation
        """
        url = f"{self.base_url}/projects/{self.project_id}/conversations/{session_id}/messages"

        response = requests.get(url, headers=self._get_headers())
        response.raise_for_status()

        data = response.json()
        if data.get("status") != "success":
            raise Exception(f"Failed to get messages: {data}")

        return data["data"]

    def update_message_reaction(
        self,
        session_id: str,
        message_id: int,
        reaction: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Update reaction for a specific message.

        Args:
            session_id: The conversation session ID
            message_id: The message ID (prompt_id)
            reaction: "liked", "disliked", or None to remove reaction

        Returns:
            dict: Updated message data with response_feedback

        Example response:
        {
            "status": "success",
            "data": {
                "id": 1,
                "user_query": "Hello",
                "openai_response": "Hi!",
                "response_feedback": {
                    "created_at": "2024-08-27T21:07:20.000000Z",
                    "updated_at": "2024-08-27T21:07:20.000000Z",
                    "user_id": 1,
                    "reaction": "liked"
                }
            }
        }

        Raises:
            ValueError: If reaction is not "liked", "disliked", or None
            requests.HTTPError: If API request fails
        """
        # Validate reaction value
        if reaction not in ["liked", "disliked", None]:
            raise ValueError(f"Invalid reaction value: {reaction}. Must be 'liked', 'disliked', or None")

        url = f"{self.base_url}/projects/{self.project_id}/conversations/{session_id}/messages/{message_id}/feedback"

        payload = {"reaction": reaction}

        response = requests.put(
            url,
            headers=self._get_headers(),
            json=payload
        )
        response.raise_for_status()

        data = response.json()
        if data.get("status") != "success":
            raise Exception(f"Failed to update reaction: {data}")

        return data["data"]

    def get_citation(self, citation_id: int) -> Dict[str, Any]:
        """
        Fetch citation details by ID.

        Args:
            citation_id: Citation identifier

        Returns:
            dict: Citation data with url, title, description, image

        Raises:
            ValueError: If citation_id is invalid
            Exception: If API request fails
        """
        if not isinstance(citation_id, int) or citation_id <= 0:
            raise ValueError(f"Invalid citation ID: {citation_id}")

        url = f"{self.base_url}/projects/{self.project_id}/citations/{citation_id}"

        response = requests.get(
            url,
            headers=self._get_headers()
        )
        response.raise_for_status()

        data = response.json()
        if data.get("status") != "success":
            raise Exception(f"Failed to fetch citation: {data}")

        return data["data"]
